### ISSUE TITLE
Merges take longer to complete over time

### ISSUE THREAD

**Issue by MonsieurNicolas** [stellar/stellar-core#2128]
## Summary

Looks like over time merges are taking longer.

Observed on nodes that watch pubnet *and* publish.

### Kernel
Linux 3.13.0-139-generic #188-Ubuntu SMP Tue Jan 9 14:43:09 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

### SKU
AWS m3.large

vCPUs: 2
model name      : Intel(R) Xeon(R) CPU E5-2670 v2 @ 2.50GHz

RAM: 8GB

Drive: EBS

## Description of the problem

As a point of comparison, here is how CPU utilization normally looks like:
![image](https://user-images.githubusercontent.com/4213763/58892992-19602280-86a4-11e9-943d-84c140ae1abc.png)

but after a few days, it degrades into:
![image](https://user-images.githubusercontent.com/4213763/58893021-2a109880-86a4-11e9-905e-a11cb4ccc5fa.png)

Restarting the machine causes the merges to complete fast again (here is a graph showing both right before and after a restart):
![image](https://user-images.githubusercontent.com/4213763/58898145-c5a70680-86ae-11e9-90f1-7262ab6b22cc.png)

### Potential root cause

It looks like `vruntime` is very high for those threads, expected as all they do is compute `SHA256`.

According to https://www.kernel.org/doc/Documentation/scheduler/sched-design-CFS.txt processes are sorted by `vruntime`, and there doesn't seem to be a decaying function of any sort: I *suspect* that over time, `vruntime` just gets so big that any other process on the machine will prempt merging.
What is confusing to me is that it doesn't look like the machine has high CPU or anything... so there might be some other throttling at play here.

Here is a diff (before/after) restart taken from `/proc/pid/sched`

```
< se.exec_start                                :   24483426539.404075
< se.vruntime                                  :     135500227.539266
< se.sum_exec_runtime                          :       6469705.584427
---
> se.exec_start                                :   24481968085.667081
> se.vruntime                                  :        114133.444487
> se.sum_exec_runtime                          :            61.248399
7c7
< se.statistics.sleep_start                    :   24483426539.404075
---
> se.statistics.sleep_start                    :   24481968085.667081
9,18c9,18
< se.statistics.sleep_max                      :         48048.484286
< se.statistics.block_max                      :          3590.918983
< se.statistics.exec_max                       :             4.418723
< se.statistics.slice_max                      :           635.302620
< se.statistics.wait_max                       :            56.031742
< se.statistics.wait_sum                       :       2987576.124842
< se.statistics.wait_count                     :              1971004
< se.statistics.iowait_sum                     :         65030.711157
< se.statistics.iowait_count                   :                 3701
< se.nr_migrations                             :                17982
---
> se.statistics.sleep_max                      :         36793.094570
> se.statistics.block_max                      :             0.042922
> se.statistics.exec_max                       :             4.000239
> se.statistics.slice_max                      :             2.275626
> se.statistics.wait_max                       :             4.028972
> se.statistics.wait_sum                       :            15.890139
> se.statistics.wait_count                     :                   79
> se.statistics.iowait_sum                     :             0.000000
> se.statistics.iowait_count                   :                    0
> se.nr_migrations                             :                   11
21,30c21,30
< se.statistics.nr_failed_migrations_running   :                15896
< se.statistics.nr_failed_migrations_hot       :                18227
< se.statistics.nr_forced_migrations           :                 3617
< se.statistics.nr_wakeups                     :               196647
< se.statistics.nr_wakeups_sync                :               172692
< se.statistics.nr_wakeups_migrate             :                 7542
< se.statistics.nr_wakeups_local               :                20343
< se.statistics.nr_wakeups_remote              :               176304
< se.statistics.nr_wakeups_affine              :                 3146
< se.statistics.nr_wakeups_affine_attempts     :                14686
---
> se.statistics.nr_failed_migrations_running   :                    2
> se.statistics.nr_failed_migrations_hot       :                    1
> se.statistics.nr_forced_migrations           :                    0
> se.statistics.nr_wakeups                     :                   48
> se.statistics.nr_wakeups_sync                :                    0
> se.statistics.nr_wakeups_migrate             :                    9
> se.statistics.nr_wakeups_local               :                   15
> se.statistics.nr_wakeups_remote              :                   33
> se.statistics.nr_wakeups_affine              :                    2
> se.statistics.nr_wakeups_affine_attempts     :                   26
33,37c33,37
< avg_atom                                     :             3.300654
< avg_per_cpu                                  :           359.787875
< nr_switches                                  :              1960128
< nr_voluntary_switches                        :               196637
< nr_involuntary_switches                      :              1763491
---
> avg_atom                                     :             0.805899
> avg_per_cpu                                  :             5.568036
> nr_switches                                  :                   76
> nr_voluntary_switches                        :                   49
> nr_involuntary_switches                      :                   27
39,42c39,42
< se.avg.runnable_avg_sum                      :                 5352
< se.avg.runnable_avg_period                   :                47726
< se.avg.load_avg_contrib                      :                   37
< se.avg.decay_count                           :          23349215069
---
> se.avg.runnable_avg_sum                      :                 2652
> se.avg.runnable_avg_period                   :                47712
> se.avg.load_avg_contrib                      :                   15
> se.avg.decay_count                           :          23347824179
45c45
< clock-delta                                  :                  116
---
> clock-delta                                  :                  168
```

**Comment by graydon** [comment]
After a little sniffing around a node experiencing this I paused to look more carefully at the actual times of the events, and .. I don't actually think these are _merges_ taking longer. I think they're background work having to do with _history publishing_:

![Screen Shot 2019-06-20 at 12 35 08 PM](https://user-images.githubusercontent.com/14097/59876409-6d336280-9358-11e9-9eb1-2de946b11cc9.png)

I'm going to look at doing some synthetic history work-publishing tests, see if maybe there's some state leak in the subprocess management code.

**Comment by graydon** [comment]
Next-most salient observation (pushing the pendulum of likelihood back towards merges as such): the ledger at which this happens (0x0173ffff) is one at which an L8->L9 merge happens, thus there are 9 merges running simultaneously, a condition which occurs about once every 3 days (which is the frequency we see this triggering at).

Why is 9 the salient boundary? Unclear. But a guess: 9 + 1 (the publication thread) = 10 which is the number of worker threads we set the default to in this version (v11.0), which is the version at which the problem started. So, hypothesis: there's something about "activating all the worker threads at once" that seems to change things?

**Comment by graydon** [comment]
I think I have an explanation now, after much digging, that fits the evidence I've gathered (including log analysis that show the load is coming not from slow merges but unexpectedly-frequent _repeated_ merges, especially large ones that should only happen every few days, starting to happen every 5 minutes, i.e. on publication events!)

Theory: back in January, commit 628e8a77bdc13d2876b1f57ddff3fa008aa3ed1b (in PR https://github.com/stellar/stellar-core/pull/1895) accidentally happened to remove the only place in the normal ledger-close execution path where a set of _live_ `FutureBuckets` are handed off directly from their initial merge `std::futures` to the `ResolveSnapshotWork` class in the publish system, that resolves them. These resolve instantly since their merge is already done and in-memory / connected to the `std::future` that the `merge` ran inside.

Instead, after #1895 history publishing always bounces its HAS off the database. This alone isn't a _huge_ problem, except that we unfortunately fail to _resolve_ the `FutureBuckets` (even just opportunistically / non-blocking -- locking those with `std::future` outputs into `FutureBucket::FB_LIVE_OUTPUT` for serialization) before we write them to an HAS in the publish queue.

(Those who may remember we do this opportunistic resolution somewhere are remembering correctly: we do it just before we write the current ledger state, but that's a separate `FutureBucket` copy and happens later in the commit path anyways. Every copy of a `FutureBucket` has its own state variable.)

Not having the `FutureBucket`s even opportunistically resolved in the publish queue HASs means that when we pull a HAS back out of the publish queue, we _reconstitute_ its `FutureBucket`s from hashes and they no longer have `std::future`s with merge-results attached to them, so resolving them for publication causes us to re-fetch the inputs from the `BucketManager` and re-run the merges. This is all working _by design_ for buckets that weren't ready at the time of enqueueing; the only problem is we failed to notice all the buckets that _were_ ready with outputs at the time of enqueueing. So we redo all those merges we only serialized the inputs for.

This makes for a CPU spike at publish time, and it will get bigger as the process runs: whatever the oldest merge that's been run (and been stored in `FutureBucket::HASH_INPUTS` state) in the publish queue, during the process lifetime, will re-run at publish time. We notice the L7->L8 step because it happens after 3 days and is relatively expensive. There'll be another such step at the first L8->L9 merge after 15 more days but I think we reboot the nodes more often than that.

This bug got into v10.2.0, and caused a small but ignored CPU spike when we deployed v10.2.0 on Feb 12-13. If you look at the graph zoomed a ways out, there's definitely a step function in CPU usage there. But we were running all the merges on high priority threads and it was small enough to be ignorable.

v11.0.0 pushed merges to low-priority background threads, which makes the spikes more noticeable (also: purple) on the dashboard. There were no other _errors_ introduced with v11.0.0, just this fact amplified the error from February.

Anyway, short story long: I _think_ that fixing this is just a matter of a one-line call to `has.resolveAnyReadyFutures()` before queueing the HAS. I am not certain but it fits all the facts, and is worth trying!
